### ISSUE TITLE
Undetected type guard implemented by a membership operator on a list

### ISSUE THREAD

**Issue by CtrlZmaster** [python/mypy#15959]
**Bug Report**

Extra return-value error is generated when a type guard is implemented as a membership check in a list. This shouldn't produce an error.

**To Reproduce**
Mypy Playground: https://mypy-play.net/?mypy=latest&python=3.11&gist=c61cac6dc4e5b54cffebb15012af0939
```python
from typing import Optional

def str_to_int(num: str) -> Optional[int]:
    try:
        return int(num) 
    except ValueError:
        return None
    
def str_to_int_nonzero(num: str) -> int:
    retval = str_to_int(num)
    if retval in [0, None]:
        raise ValueError(f"'{num}' is not a non-zero integer")
    return retval
```

**Actual Behavior**
Mypy output:
```
example.py:13: error: Incompatible return value type (got "int | None", expected "int")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```
I believe that this is incorrect, if the `str_to_int` function returns `None`, the membership check will catch it. It seems that mypy completely ignores the membership checks, even using `if something in [None]` instead of `if something is None` generates an error. Membership check does an identity check first and then an equality check, so I think that this is a valid way of implementing a type guard for narrowing `Optional[int]` to `int`. (At least in CPython - `in` operator on a list uses [list_contains](https://github.com/python/cpython/blob/main/Objects/listobject.c#L443) which uses [this comparison](https://github.com/python/cpython/blob/main/Objects/object.c#L846) - "Quick result when objects are the same. Guarantees that identity implies equality".) 

**Your Environment**
- Mypy version used: mypy 1.5.1 (compiled: yes)
- Mypy command-line flags: no flags
- Mypy configuration options from `mypy.ini` (and other config files): no config used
- Python version used: Python 3.10.12

**Comment by AlexWaygood** [comment]
Duplicate of #10977
